### ISSUE TITLE
Unsupported operand types

### ISSUE THREAD

**Issue by P147x** [guzzle/guzzle#2016]
Hello,
I'm actually trying to use Auth0 integration on a Prestashop module, but after installed the integration with Composer, i've this error :

[PrestaShop] Fatal error in module file: /var/www/html/modules/prestashop/vendor/guzzlehttp/guzzle/src/Client.php:236
Unsupported operand types

Is that a bug ? It's actually stucking me.
Thanks in advance.

Lucas

**Comment by sagikazarmark** [comment]
What is your Guzzle version? What is your PHP version?

A bit more detailed stack trace would be nice, because without that you won't be able to find the usage of the library and without input parameters we cannot reproduce the issue.

Looking at the latest code on the aformentioned line there is an array merging operation which awaits two arrays. Maybe there is an uncompatible input there.

**Comment by P147x** [comment]
I'm sorry, it's actually the first time i'm asking help here for something.
My PHP version is 7.1.14 (running on a Docker), and it's actually installed with Auth0 package (Composer) so i don't have the version but i guess it should be the last one..

About the logs, it's all i got.

**Comment by sagikazarmark** [comment]
> I'm sorry, it's actually the first time i'm asking help here for something.

No problem, just a friendly reminder. ;)

> (Composer) so i don't have the version but i guess it should be the last one..

You can check the version in the composer.lock file

**Comment by P147x** [comment]
I'll keep that in my mind for the future, thanks :)

I'm using guzzlehttp 6.3.0

**Comment by sagikazarmark** [comment]
Without the actual stack trace it's going to be hard to identify the problem. Can you debug the code?

If you have no debugging tools at hand, at least could you go to that line and var_dump the two operands (poor mans debug)?

**Comment by P147x** [comment]
I've only the code of the guzzle library on my folder. The only line of code i've called is 
"require_once(dirname(__FILE__) . '/vendor/autoload.php');"

It concern the operand "$result = $options + $defaults;"

Here's the complete function affiliated :
`private function prepareDefaults($options)
    {
        $defaults = $this->config;

        if (!empty($defaults['headers'])) {
            // Default headers are only added if they are not present.
            $defaults['_conditional'] = $defaults['headers'];
            unset($defaults['headers']);
        }

        // Special handling for headers is required as they are added as
        // conditional headers and as headers passed to a request ctor.
        if (array_key_exists('headers', $options)) {
            // Allows default headers to be unset.
            if ($options['headers'] === null) {
                $defaults['_conditional'] = null;
                unset($options['headers']);
            } elseif (!is_array($options['headers'])) {
                throw new \InvalidArgumentException('headers must be an array');
            }
        }

        // Shallow merge defaults underneath options.
        $result = $options + $defaults;

        // Remove null values.
        foreach ($result as $k => $v) {
            if ($v === null) {
                unset($result[$k]);
            }
        }

        return $result;
    }`

**Comment by sagikazarmark** [comment]
Yup, that's it. This line in particular:

`$result = $options + $defaults;`

Can you dump those two variables before merging them?

**Comment by P147x** [comment]
Here's what i got.
The first (string) is the $options value, and the second one is $defaults.

`=========

string(38) "1var/www/html/app/cache/dev/cacert.pem"
array(9) { ["base_url"]=> string(33) "https://api-addons.prestashop.com" ["defaults"]=> array(1) { ["timeout"]=> string(3) "5.0" } ["handler"]=> object(GuzzleHttp\HandlerStack)#393 (3) { ["handler":"GuzzleHttp\HandlerStack":private]=> object(Closure)#446 (2) { ["static"]=> array(2) { ["default"]=> object(Closure)#416 (2) { ["static"]=> array(2) { ["default"]=> object(GuzzleHttp\Handler\CurlMultiHandler)#457 (5) { ["factory":"GuzzleHttp\Handler\CurlMultiHandler":private]=> object(GuzzleHttp\Handler\CurlFactory)#459 (2) { ["handles":"GuzzleHttp\Handler\CurlFactory":private]=> array(0) { } ["maxHandles":"GuzzleHttp\Handler\CurlFactory":private]=> int(50) } ["selectTimeout":"GuzzleHttp\Handler\CurlMultiHandler":private]=> int(1) ["active":"GuzzleHttp\Handler\CurlMultiHandler":private]=> NULL ["handles":"GuzzleHttp\Handler\CurlMultiHandler":private]=> array(0) { } ["delays":"GuzzleHttp\Handler\CurlMultiHandler":private]=> array(0) { } } ["sync"]=> object(GuzzleHttp\Handler\CurlHandler)#454 (1) { ["factory":"GuzzleHttp\Handler\CurlHandler":private]=> object(GuzzleHttp\Handler\CurlFactory)#389 (2) { ["handles":"GuzzleHttp\Handler\CurlFactory":private]=> array(0) { } ["maxHandles":"GuzzleHttp\Handler\CurlFactory":private]=> int(3) } } } ["parameter"]=> array(2) { ["$request"]=> string(10) "" ["$options"]=> string(10) "" } } ["streaming"]=> object(GuzzleHttp\Handler\StreamHandler)#394 (1) { ["lastHeaders":"GuzzleHttp\Handler\StreamHandler":private]=> array(0) { } } } ["parameter"]=> array(2) { ["$request"]=> string(10) "" ["$options"]=> string(10) "" } } ["stack":"GuzzleHttp\HandlerStack":private]=> array(4) { [0]=> array(2) { [0]=> object(Closure)#464 (1) { ["parameter"]=> array(1) { ["$handler"]=> string(10) "" } } [1]=> string(11) "http_errors" } [1]=> array(2) { [0]=> object(Closure)#469 (1) { ["parameter"]=> array(1) { ["$handler"]=> string(10) "" } } [1]=> string(15) "allow_redirects" } [2]=> array(2) { [0]=> object(Closure)#470 (1) { ["parameter"]=> array(1) { ["$handler"]=> string(10) "" } } [1]=> string(7) "cookies" } [3]=> array(2) { [0]=> object(Closure)#423 (1) { ["parameter"]=> array(1) { ["$handler"]=> string(10) "" } } [1]=> string(12) "prepare_body" } } ["cached":"GuzzleHttp\HandlerStack":private]=> NULL } ["allow_redirects"]=> array(5) { ["max"]=> int(5) ["protocols"]=> array(2) { [0]=> string(4) "http" [1]=> string(5) "https" } ["strict"]=> bool(false) ["referer"]=> bool(false) ["track_redirects"]=> bool(false) } ["http_errors"]=> bool(true) ["decode_content"]=> bool(true) ["verify"]=> bool(true) ["cookies"]=> bool(false) ["_conditional"]=> array(2) { ["Accept"]=> string(16) "application/json" ["User-Agent"]=> string(39) "GuzzleHttp/6.2.1 curl/7.38.0 PHP/5.6.33" } }

========`

(Sorry for the late answer)

**Comment by sagikazarmark** [comment]
Well, you will have to debug how that value reached the prepare method. It's provided by some external caller.

**Comment by murilohns** [comment]
Hello @P147x, did you solved that problem?

**Comment by soldierm** [comment]
Hello @P147x, did you solved that problem?

**Comment by P147x** [comment]
Hello guys, sorry for the last answer.
Its been a time, and i forgot a bit. I think the problem was the Auth0 integration who was currently using guzzle, maybe in a wrong way. Are you having a similar issue ?

**Comment by weinasi** [comment]
I have  a similar issue.This is my error trace:
[2020-03-04 00:45:17] test.ERROR: Unsupported operand types {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Unsupported operand types at /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/guzzlehttp/guzzle/src/Client.php:320)
[stacktrace]
#0 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/guzzlehttp/guzzle/src/Client.php(149): GuzzleHttp\\Client->prepareDefaults(Array)
#1 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/guzzlehttp/guzzle/src/Client.php(183): GuzzleHttp\\Client->requestAsync('shouldReport', Array, Array)
#2 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/guzzlehttp/guzzle/src/Client.php(96): GuzzleHttp\\Client->request('shouldReport', Array, Array)
#3 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/guzzlehttp/guzzle/src/Client.php(63): GuzzleHttp\\Client->__call('shouldReport', Array)
#4 /data/wwwroot/test.linde.smartlink.aidong.ai/app/Http/Controllers/Carrier/CarrierDeviceTrait.php(63): GuzzleHttp\\Client->__construct(Array)
#5 /data/wwwroot/test.linde.smartlink.aidong.ai/app/Http/Controllers/Carrier/CarrierDeviceTrait.php(578): App\\Jobs\\DeviceMseJob::getGuzzleHttpBySync('device/getCarri...', 'deviceId=866202...')
#6 /data/wwwroot/test.linde.smartlink.aidong.ai/app/Jobs/DeviceMseJob.php(117): App\\Jobs\\DeviceMseJob->getCarrierInfoByDeviceId('866202001201904')
#7 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(88): App\\Jobs\\DeviceMseJob->fire(Object(Freyo\\LaravelQueueCMQ\\Queue\\Jobs\\CMQJob), Array)
#8 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/freyo/laravel-queue-cmq/src/Queue/Jobs/CMQJob.php(81): Illuminate\\Queue\\Jobs\\Job->fire()
--
#22 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/symfony/console/Application.php(1012): Illuminate\\Console\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#23 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/symfony/console/Application.php(272): Symfony\\Component\\Console\\Application->doRunCommand(Object(Illuminate\\Queue\\Console\\WorkCommand), Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#24 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/symfony/console/Application.php(148): Symfony\\Component\\Console\\Application->doRun(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#25 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/laravel/framework/src/Illuminate/Console/Application.php(90): Symfony\\Component\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#26 /data/wwwroot/test.linde.smartlink.aidong.ai/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(133): Illuminate\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#27 /data/wwwroot/test.linde.smartlink.aidong.ai/artisan(35): Illuminate\\Foundation\\Console\\Kernel->handle(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#28 {main}
"}

**Comment by weinasi** [comment]
my $options value and $defaults value:
options:{"synchronous":true} defaults:{"handler":{},"base_uri":{},"allow_redirects":{"max":5,"protocols":["http","https"],"strict":false,"referer":false,"track_redirects":false},"http_errors":true,"decode_content":true,"verify":true,"cookies":false,"idn_conversion":true,"_conditional":{"User-Agent":"GuzzleHttp\/6.5.1 curl\/7.56.0 PHP\/7.1.32"}}

**Comment by weinasi** [comment]
this is my trace:
[2020-03-04 00:45:17] test.ERROR: Unsupported operand types {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Unsupported operand types at /data/wwwroot/test_snow/vendor/guzzlehttp/guzzle/src/Client.php:320)
[stacktrace]
#0 /data/wwwroot/test_snow/vendor/guzzlehttp/guzzle/src/Client.php(149): GuzzleHttp\\Client->prepareDefaults(Array)
#1 /data/wwwroot/test_snow/vendor/guzzlehttp/guzzle/src/Client.php(183): GuzzleHttp\\Client->requestAsync('shouldReport', Array, Array)
#2 /data/wwwroot/test_snow/vendor/guzzlehttp/guzzle/src/Client.php(96): GuzzleHttp\\Client->request('shouldReport', Array, Array)
#3 /data/wwwroot/test_snow/vendor/guzzlehttp/guzzle/src/Client.php(63): GuzzleHttp\\Client->__call('shouldReport', Array)

**Comment by connor11528** [comment]
also having issue with `Unsupported operand types`. Looks like `$options` and `$defaults` are both array types, should we really be merging them with a `+`?

```
// Shallow merge defaults underneath options.
$result = $options + $defaults;
```

**Comment by mfn** [comment]
> should we really be merging them with a `+`?

This is perfectly fine and well documented at https://www.php.net/manual/en/language.operators.array.php
>  The + operator returns the right-hand array appended to the left-hand array; for keys that exist in both arrays, the elements from the left-hand array will be used, and the matching elements from the right-hand array will be ignored. 

Or in plain english: it's the most common way to have a user-provided array and a defaults-array and produce a result where all the defaults _missing_ in the user-provided are filled in. The method is poignantly called `prepareDefaults` for that reason.

Everyone having the problem: the problem is that something _outside_ Guzzle passes data in where either operand is not an array anymore, as is evident by https://github.com/guzzle/guzzle/issues/2016#issuecomment-368227484

**Comment by Nyholm** [comment]
Thank you @mfn

**Comment by FrancoisPrigent** [comment]
It happens what you don't call the Guzzle client construct method.

**Comment by carltondickson** [comment]
This is my stack trace for a similar issue with `PHP7.4` and Guzzle `7.2.0`

```
#42 /vendor/guzzlehttp/guzzle/src/Client.php(304): GuzzleHttp\Client::prepareDefaults
#41 /vendor/guzzlehttp/guzzle/src/Client.php(154): GuzzleHttp\Client::requestAsync
#40 /vendor/guzzlehttp/guzzle/src/Client.php(187): GuzzleHttp\Client::request
#39 /vendor/guzzlehttp/guzzle/src/Client.php(92): GuzzleHttp\Client::__call
#38 /vendor/guzzlehttp/guzzle/src/Client.php(57): GuzzleHttp\Client::__construct
```

and part of the code

```
            $http = new Client([
                'verify' => App::environment('local') ? false : true,
            ]);
            $response = $http->get($url, [
                'headers' => $headers,
                'query'   => $query,
            ]);
```

Noticed that when running this locally I end up in the `get` method inside of `vendor/guzzlehttp/guzzle/src/ClientTrait.php` method instead of the `__call` method where the problem merging options and defaults seems to trigger.

This piece of code runs as part of a job and the server also has New Relic extension installed on it...not sure if this is the problem but found this (https://discuss.newrelic.com/t/php-guzzle-errors-in-php-agent-9-0/91477/22)

I'm trying to figure out how this happens and will update this post if I can find out why.

**Comment by mfn** [comment]
As to the reference with the newrelic forum, be careful, the last comment says:
> from time to time it somehow breaks, which causes all guzzle related jobs to fail until the newrelic daemon is restarted

Which is a clear indicator it's unrelated to Guzzle.

> `/vendor/guzzlehttp/guzzle/src/Client.php(304)`

This is exactly this line:
```php
        $result = $options + $defaults;
```
Either `$options` or `$defaults` is not an array (anymore).

I suggest to temporarily add debugging code before this line to witness what those two values are and check it out once you hit problem again.

**Comment by carltondickson** [comment]
Running locally, my debugger stops in the `get` method as expected but on the remote server instead of calling `get` it tries to find this method via `__call` given the stack trace above.

Still digging, but what could cause `Client` to execute the magic method `__call` instead of the `get` method in `vendor/guzzlehttp/guzzle/src/ClientTrait.php`.

**Comment by mfn** [comment]
> but what could cause `Client` to execute the magic method `__call` instead of the `get` method

Oh, thanks for pointing this out again! You already mentioned it but I missed it. This is _very_ strange.

- `Client.php(57)` -> https://github.com/guzzle/guzzle/blob/de6f1e58e735754b888649495ed4cb9ae3b19589/src/Client.php#L57 (not so useful)
- `Client.php(92)` -> https://github.com/guzzle/guzzle/blob/de6f1e58e735754b888649495ed4cb9ae3b19589/src/Client.php#L92

What perplexes me, it looks like there something missing: how could it get there? 

> but on the remote server instead of calling `get`

Ah, _that_ `get` method from the trait 😅 Man, this naming is confusing => https://github.com/guzzle/guzzle/blob/de6f1e58e735754b888649495ed4cb9ae3b19589/src/ClientTrait.php#L42

Is there a problem with code being not up2date, weird caching?

Looks like as if the `ClientTrait` is not applied? 🤔 Do you source versions match?

**Comment by carltondickson** [comment]
This error started occuring after upgrading a project from Laravel 5.x to 8 (Guzzle `6.5.5` to `7.2.0`).
Will see if there is any sort of caching going on, it definitely feels like something is still trying to use some old code where this trait doesn't exist. We're running the project on Heroku so anything is possible once the code is running on one of their dynos.

The composer log on deploy shows this

```
- Installing guzzlehttp/guzzle (7.2.0): Extracting archive
```

Checked the vendor directory on the server and can see that this trait exists and looks correct in the sense that the `get` method is there

**Comment by sblawrie** [comment]
I'm also having this issue on Laravel 8, Guzzle 7.3. I use New Relic, so I think it's probably related to that.

**Comment by LasseRafn** [comment]
I'm seeing this after installing New Relic, never before. I suppose its new relic related

**Comment by fhferreira** [comment]
anyone solve it?

**Comment by GrahamCampbell** [comment]
Upgrade to the latest version. You may need ton upgrade both local and global.
